lib/urlapi.c

```diff
@@ -1307,6 +1307,178 @@ CURLU *curl_url_dup(const CURLU *in)
   return NULL;
 }
 
+static char *normalize_path(const char *path)
+{
+  /* split the path per slash, URL decode + encode, then put together again */
+  size_t len = strlen(path);
+  char *sl;
+  struct dynbuf dupe;
+  Curl_dyn_init(&dupe, len * 3);
+
+  do {
+    char *opath;
+    char *npath;
+    int olen;
+    size_t partlen;
+    sl = memchr(path, '/', len);
+    partlen = sl ? (size_t)(sl - path) : len;
+
+    if(partlen) {
+      /* First URL decode the part */
+      opath = curl_easy_unescape(NULL, path, (int)partlen, &olen);
+      if(!opath) {
+        Curl_dyn_free(&dupe);
+        return NULL;
+      }
+
+      /* Then URL encode it again */
+      npath = curl_easy_escape(NULL, opath, olen);
+      curl_free(opath);
+      if(!npath) {
+        Curl_dyn_free(&dupe);
+        return NULL;
+      }
+
+      if(Curl_dyn_add(&dupe, npath) || (sl && Curl_dyn_addn(&dupe, "/", 1))) {
+        curl_free(npath);
+        return NULL;
+      }
+      curl_free(npath);
+    }
+    else if(sl) {
+      /* zero length part but a slash */
+      if(Curl_dyn_addn(&dupe, "/", 1))
+        return NULL;
+    }
+    else {
+      /* no part, no slash */
+      break;
+    }
+
+    if(sl) {
+      path = sl + 1;
+      len -= partlen + 1;
+    }
+
+  } while(sl);
+
+  return Curl_dyn_ptr(&dupe);
+}
+
+/* leave these as-is in query parts */
+#define ISQUERYOK(x) (((x) == ';') || ((x) == '&') || ((x) == '=') ||   \
+                      ((x) == ':') || ((x) == '/') || ((x) == '?') ||   \
+                      ((x) == '+'))
```

Comment:
What is the reason for choosing these specific characters, and leaving off `@`, `!`, `$`, `'`, `(`, `)`, `*`, `[`, `]`, and `,`?  I know that `;`, `&`, and `=` have special meanings, but was not aware of the others.  I think this should be included in the documentation, so that users can understand the exact algorithm libcurl implements without reading the source code.

Other characters are either forbidden in URLs (must be encoded), unreserved (should be decoded), or `#` (which cannot appear unencoded in a query part because it starts the fragment).